File: java/src/main/java/com/dokmatiq/docgen/client/TemplatesClient.java
package com.dokmatiq.docgen.client;

import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.internal.Transport;

import java.nio.file.Path;
import java.util.List;
import java.util.Map;

/** Client for template management endpoints. */
public class TemplatesClient {
    private final Transport transport;

    public TemplatesClient(Transport transport) {
        this.transport = transport;
    }

    /** Upload a template file (ODT/DOCX). */
    public Map<?, ?> upload(Path file) {
        return transport.upload("/api/templates", "file",
                FileUtils.readBytes(file), FileUtils.detectFilename(file), null, Map.class);
    }

    /** Upload a template from bytes. */
    public Map<?, ?> upload(byte[] data, String fileName) {
        return transport.upload("/api/templates", "file", data, fileName, null, Map.class);
    }

    /** List all uploaded templates. */
    public List<Map> list() {
        return transport.requestList("GET", "/api/templates", Map.class);
    }

    /** Delete a template by name. */
    public void delete(String name) {
        transport.delete("/api/templates/" + name);
    }
}

File: dotnet/src/DocGen/Clients/TemplatesClient.cs
using Dokmatiq.DocGen.Internal;

namespace Dokmatiq.DocGen.Clients;

/// <summary>
/// Client for template management endpoints.
/// </summary>
public sealed class TemplatesClient
{
    private readonly Transport _transport;

    internal TemplatesClient(Transport transport) => _transport = transport;

    /// <summary>Upload a template file (ODT/DOCX).</summary>
    public Dictionary<string, object> Upload(string filePath)
        => _transport.Upload<Dictionary<string, object>>("/api/templates", "file",
            FileUtils.ReadBytes(filePath), FileUtils.DetectFilename(filePath));

    /// <summary>Upload a template from bytes.</summary>
    public Dictionary<string, object> Upload(byte[] data, string fileName)
        => _transport.Upload<Dictionary<string, object>>("/api/templates", "file", data, fileName);

    /// <summary>List all uploaded templates.</summary>
    public List<Dictionary<string, object>> List()
        => _transport.RequestList<Dictionary<string, object>>(HttpMethod.Get, "/api/templates");

    /// <summary>Delete a template by name.</summary>
    public void Delete(string name) => _transport.Delete($"/api/templates/{name}");
}
